File: lena-config/src/main/java/org/riversoforion/lena/config/CompositeConfigurationSource.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class CompositeConfigurationSource implements ConfigurationSource {

    private final List<NamespacedWrapper> configurationSources;

    protected CompositeConfigurationSource(ConfigurationSource... configurationSources) {

        this.configurationSources = Arrays.stream(configurationSources).map(CompositeConfigurationSource::namespaced).collect(Collectors.toList());
    }

    protected CompositeConfigurationSource(Collection<ConfigurationSource> configurationSources) {

        this.configurationSources = configurationSources.stream().map(CompositeConfigurationSource::namespaced).collect(Collectors.toList());
    }

    protected CompositeConfigurationSource(Map<ConfigurationSource, Namespace> configurationSources) {

        this.configurationSources = configurationSources.entrySet().stream().map(CompositeConfigurationSource::namespaced).collect(Collectors.toList());
    }

    public static NamespacedWrapper namespaced(ConfigurationSource configurationSource, Namespace namespace) {

        if (configurationSource instanceof NamespacedWrapper namespaced) {
            return namespaced;
        }
        return new NamespacedWrapper(namespace, configurationSource);
    }

    public static NamespacedWrapper namespaced(ConfigurationSource configurationSource) {

        return namespaced(configurationSource, null);
    }

    public static NamespacedWrapper namespaced(Map.Entry<ConfigurationSource, Namespace> entry) {

        return namespaced(entry.getKey(), entry.getValue());
    }

    @Override
    public final Optional<String> getValue(Namespace namespace, Name name) {

        return orderedSources(configurationSources).map(wrapper -> wrapper.getValue(namespace, name))
                                                   .filter(Optional::isPresent)
                                                   .map(Optional::get)
                                                   .findFirst();
    }

    protected abstract Stream<NamespacedWrapper> orderedSources(List<NamespacedWrapper> sources);

    public static class NamespacedWrapper implements ConfigurationSource {

        private final Optional<Namespace> namespace;
        private final ConfigurationSource configurationSource;

        private NamespacedWrapper(Namespace namespace, ConfigurationSource configurationSource) {

            this.namespace = Optional.ofNullable(namespace);
            this.configurationSource = Objects.requireNonNull(configurationSource);
        }

        public Optional<Namespace> namespace() {

            return namespace;
        }

        public ConfigurationSource configurationSource() {

            return configurationSource;
        }

        @Override
        public Optional<String> getValue(Namespace namespace, Name name) {

            return configurationSource.getValue(this.namespace.orElse(namespace), name);
        }
    }
}

File: lena-config/src/commonMain/kotlin/org/riversoforion/lena/config/CompositeConfigurationSource.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

/**
 * A [ConfigurationSource] that queries multiple delegate sources. Subclasses control ordering
 * via [orderedSources].
 *
 * Each entry in [sources] may carry an optional [Namespace] that is prepended to lookup
 * namespaces before delegating (used for scoped sub-configurations).
 */
public abstract class CompositeConfigurationSource protected constructor(
    protected val sources: List<Pair<ConfigurationSource, Namespace?>>,
) : ConfigurationSource {

    public companion object {
        public fun namespaced(source: ConfigurationSource, namespace: Namespace): ConfigurationSource =
            object : CompositeConfigurationSource(listOf(source to namespace)) {
                override fun orderedSources(
                    all: List<Pair<ConfigurationSource, Namespace?>>
                ): List<Pair<ConfigurationSource, Namespace?>> = all
            }

        public fun namespaced(source: ConfigurationSource): ConfigurationSource =
            namespaced(source, Namespace.root())
    }

    override fun getValue(namespace: Namespace, name: Name): String? =
        orderedSources(sources).firstNotNullOfOrNull { (source, ns) ->
            val effectiveNamespace = if (ns != null && ns != Namespace.root()) {
                Namespace.parse(ns.name() + namespace.name())
            } else {
                namespace
            }
            source.getValue(effectiveNamespace, name)
        }

    protected abstract fun orderedSources(
        all: List<Pair<ConfigurationSource, Namespace?>>,
    ): List<Pair<ConfigurationSource, Namespace?>>
}
